libshpool/src/config.rs

```diff
@@ -33,90 +34,114 @@ use super::{daemon::keybindings, user};
+            ConfigWatcher::new(move || {
+                info!("reloading config");
+                let mut config = config.write().unwrap();
+                match Self::load(&config_files) {
+                    Ok(c) => {
+                        info!("new config: {:?}", c);
+                        *config = c;
+                    }
+                    Err(err) => warn!("error loading config file: {:?}", err),
                 }
-                Err(e) => warn!("config file watch err: {:?}", e),
             })
-            .context("building watcher")?;
-            watcher
-                .watch(Path::new(&watch_path), notify::RecursiveMode::NonRecursive)
-                .context("registering config file for watching")?;
-            manager.watcher = Some(Arc::new(watcher));
+            .context("building watcher")?
+        };
+        for path in config_files {
+            watcher.watch(path).context("registering config file for watching")?;
         }
+        let manager = Manager { config, watcher: Arc::new(watcher), dirs };
 
         Ok(manager)
     }
 
-    // Get the current config value.
+    /// Get the current config value.
     pub fn get(&self) -> RwLockReadGuard<'_, Config> {
         self.config.read().unwrap()
     }
-}
 
-impl std::clone::Clone for Manager {
-    fn clone(&self) -> Self {
-        Manager { config: Arc::clone(&self.config), watcher: self.watcher.as_ref().map(Arc::clone) }
+    /// Get the ProjectDirs instance created at startup.
+    #[allow(dead_code)]
+    pub fn dirs(&self) -> &ProjectDirs {
+        &self.dirs
+    }
+
+    /// Load config by merging configurations from a list of Paths.
+    ///
+    /// Paths come later in the list takes higher priority.
+    /// Merge strategy is as defined in `Config::or`.
+    fn load(config_files: impl IntoIterator<Item = impl AsRef<Path>>) -> anyhow::Result<Config> {
+        let mut config = Config::default();
+        for path in config_files {
+            let path = path.as_ref();
+            let config_str = match fs::read_to_string(path) {
+                // It is okay if the file is not there.
+                Err(e) if e.kind() == io::ErrorKind::NotFound => continue,
+                Err(e) => {
+                    warn!("error reading config file: {:?}", e);
+                    return Err(e)
+                        .with_context(|| format!("reading config toml {}", path.to_string_lossy()));
+                }
+                Ok(s) => s,
+            };
+            let new_config: Config = match toml::from_str(&config_str) {
+                Err(e) => {
+                    warn!("error parsing config file: {:?}", e);
+                    return Err(e).with_context(|| {
+                        format!("parsing config toml {}", path.to_string_lossy())
+                    });
+                }
+                Ok(c) => c,
+            };
+            config = new_config.or(config);
```

Comment:
I feel like `or` is a little deceptive since the operation is not comunative. I think we should go with a name like `merge` that does not have mathy implications which don't hold up.